0007092: Added a target_node_id column to sym_node_channel_ctl

evan-miller-jumpmind opened the pull request.

evan-miller-jumpmind committed:
0007092: Added a target_node_id column to sym_node_channel_ctl

evan-miller-jumpmind committed:
0007092: Added a target_node_id column to sym_node_channel_ctl

evan-miller-jumpmind committed:
0007092: Fixed failing unit test

pavel-jm commented on symmetric-core/src/main/java/org/jumpmind/symmetric/model/ChannelMap.java:
```diff
@@ -34,58 +35,101 @@
+    private String getChannelsAsString(String key, String targetNodeId) {
+        Set<String> channelIdSet = new HashSet<String>();
+        Map<String, Set<String>> includedChannelMap = map.get(key);
+        Map<String, Set<String>> excludedChannelMap = map.get(key.equals(CHANNELS_SUSPEND) ? CHANNELS_UNSUSPEND : CHANNELS_UNIGNORE);
```
In order to be more defensive about potential future problem, it is best to write any assumptions made about the "key" argument to this method in code? 
Let's guard against values not matching those assumptions, like... null, CHANNELS_UNIGNORE, CHANNELS_UNSUSPEND (!), etc.
This could be done by throwing an IllegalArgumentException at the very top of this method.

pavel-jm commented on symmetric-core/src/main/java/org/jumpmind/symmetric/model/ChannelMap.java:
```diff
@@ -34,58 +35,101 @@
+    public static final String CHANNELS_UNIGNORE = "Unignored-Channels";
     private String channelQueue;
-    private Map<String, Set<String>> map;
+    private final Map<String, Map<String, Set<String>>> map = new HashMap<String, Map<String, Set<String>>>();
```
When you have some time, let's meet and brainstorm on creating a new wrapper class(es)?
It would cover three main goals:
1. Keep code clean and easy to read by replacing lengthy Map<String, Set<String>> (and, also Map<String, Map<String, Set<String>>>) with a single class name.
2. Move some tedious, but must-unit-test functionality into these "wrapper" classes.
3. Simplify unit testing down to a trivial level.

There are a couple ways to implement this, so let's meet and chat about just about the class+package architecture, when you have some time?

pavel-jm commented on symmetric-core/src/main/java/org/jumpmind/symmetric/model/OutgoingBatches.java:
```diff
@@ -119,6 +120,27 @@ public List<OutgoingBatch> filterBatchesForChannels(Set<String> channels) {
+                }
+            }
+        }
+        batches.removeAll(filtered);
```
If all logic above this line were in a separate method, it would be must more readable and easy to unit-test! 
Both benefits are very valuable to the team long-term.

pavel-jm commented on symmetric-core/src/main/java/org/jumpmind/symmetric/route/ConfigurationChangedDataRouter.java:
```diff
@@ -98,6 +99,17 @@ public Set<String> routeToNodes(SimpleRouterContext routingContext, DataMetaData
                 routeNodeTables(nodeIds, columnValues, rootNetworkedNode, me, routingContext,
                         dataMetaData, possibleTargetNodes, initialLoad);
+            } else if (tableMatches(dataMetaData, TableConstants.SYM_NODE_CHANNEL_CTL)) {
+                String sourceNodeId = columnValues.get("NODE_ID");
```
From the first look the code inside the IF block appears to be a great candidate for a new method?

pavel-jm reviewed (changes requested):
Let's meet (on Tuesday?) to discuss these comments and which ones we want to prioritize for adoption?

evan-miller-jumpmind committed:
0007092: Made it possible to clear the cache while saving a sym_node_channel_ctl row

evan-miller-jumpmind committed:
0007092: Added wrapper classes for ChannelMap, handled invalid key argument, moved some code into separate methods, and included new OutgoingBatches method in unit tests

pavel-jm commented on symmetric-core/src/main/java/org/jumpmind/symmetric/model/ChannelNodesMap.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.TreeMap;
+
+public class TargetNodeMap {
+    private Map<String, Set<String>> map = new TreeMap<String, Set<String>>();
```
Could you add a note (here or at a class level) about the purpose of this collection and it's elements, please?

evan-miller-jumpmind committed:
0007092: Added comment to TargetNodeMap

evan-miller-jumpmind committed:
0007092: Removed unused createIfAbsent argument

evan-miller-jumpmind committed:
Merge branch 'release/3.17' into feature/7092_add_target_node_id_to_sym_node_channel_ctl

evan-miller-jumpmind committed:
Merge branch 'release/3.17' into feature/7092_add_target_node_id_to_sym_node_channel_ctl

evan-miller-jumpmind committed:
Merge branch 'release/3.17' into feature/7092_add_target_node_id_to_sym_node_channel_ctl

evan-miller-jumpmind committed:
0007092: Removed ALL keyword, renamed ChannelMap and related classes and methods, added methods to NodeChannels, deleted relevant sym_node_channel_ctl rows during unregistration, excluded last_extract_time from sync, and updated last_extract_time description

evan-miller-jumpmind commented:
I made the changes that we discussed:

- Removed the "ALL" keyword
- Changed `ChannelMapWrapper` -> `ChannelMap` -> `TargetNodeMap` to `NodeChannels` -> `ChannelNodesMap` and got rid of the outermost map that just had 2 entries
- Added `isBatchSuspended()` and `isBatchIgnored()` methods to `NodeChannels`
- Added "channel" to some `ChannelNodesMap` method names
- When unregistering a node, deleted all `sym_node_channel_ctl` rows that contain its node ID
- Excluded `last_extract_time` column from sync
- Updated `last_extract_time` column description

Here are the remaining questions that I have, which we can discuss here or in a meeting:

1. The `target_node_id` column isn't needed when keeping track of the `last_extract_time`. Previously, I was using a row with `target_node_id='ALL'` to keep track of the `last_extract_time`. Now that "ALL" has been removed, I'm using a row where `target_node_id=node_id`, and also excluding such rows from routing because they have no other use. Is that okay?
2. What should we do about the new `target_node_id` column during an upgrade from a pre-3.17 version?

pavel-jm commented:
RE:
>     1. The `target_node_id` column isn't needed when keeping track of the `last_extract_time`. Previously, I was using a row with `target_node_id='ALL'` to keep track of the `last_extract_time`. Now that "ALL" has been removed, I'm using a row where `target_node_id=node_id`, and also excluding such rows from routing because they have no other use. Is that okay?

Without looking at the actual SQL and runtime details, I'm thinking that both nodes should be in the query ?
where node_id=<current_node_id> and target_node_id=<target_node_id>

My concern is that using only the target_node_id column to filter is not specific enough, especially in a 3-tier system.

Maybe we should have a micro-demo in the next meeting to better understand it?

pavel-jm commented:
RE: 
>     2. What should we do about the new `target_node_id` column during an upgrade from a pre-3.17 version?

My proposal is to delete all records from the SYM_NODE_CHANNEL_CTL and list this as a breaking change in documentation. Hardly anyone uses this table, so spending a lot of time on this seems wasteful?

evan-miller-jumpmind committed:
0007092: Changed how last_extract_time is tracked, deleted from sym_node_channel_ctl during an upgrade, and documented sym_node_channel_ctl improvements in What's New section

pavel-jm commented on symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/DataExtractorService.java:
```diff
@@ -558,10 +558,10 @@ protected List<OutgoingBatch> extract(final ProcessInfo extractInfo, final Node
             OutgoingBatch currentBatch = null;
             ExecutorService executor = null;
+            Node sourceNode = nodeService.findIdentity();
             try {
```
I've noticed the call to nodeService.findIdentity() had moved to be outside the try{}catch block.
How does the risk landscape change with this, please?
Hoping this is arrangement is "easier" for support team to troubleshoot?

evan-miller-jumpmind replied:
Looking at the implementation of `NodeService.findIdentity()`, I don't see a risk of it throwing an exception. It may have to run a SQL query if the identity isn't cached, but it catches a `SqlException` if it occurs.
